DOCS-10594-jms-activeMQ-kt (#1416)
mulesoft/docs-connectors: jms/1.7/modules/ROOT/pages/jms-activemq-configuration.adoc

> = Configuring an ActiveMQ Connection

Reviewer: edit:
Configure Message Redelivery
Configure Trusted Packages

easy to start building an application without configuring a connection against an external broker.
The in-memory broker is required to configure both the *ActiveMQ KahaDB* and *ActiveMQ Broker* libraries. The default URL is:

`vm://localhost?broker.persistent=false&broker.useJmx=false`

To configure the in-memory broker in Studio, follow these steps:

. In Studio, open the *JMS Config* global element window.
. In the *Connection Factory* section, set the *Factory configuration* field to *Edit inline*.
. Set the *Broker url* field to the address of the broker to connect, for example, `vm://localhost?broker.persistent=false&broker.useJmx=false` .
. Click *OK*.

.JMS Connector In-Memory Broker Configuration
image::jms-activemq-broker.png[Set the Broker url field to the address of the broker to connect]

In the XML editor, the `brokerUrl` configuration looks like this:

[source,xml,linenums]
----
<jms:config name="JMS_Config" doc:name="JMS Config">
		<jms:active-mq-connection >
			<jms:factory-configuration brokerUrl="vm://localhost?broker.persistent=false&amp;broker.useJmx=false"/>
		</jms:active-mq-connection>
	</jms:config>
----

> == Configure Message Redelivery

Reviewer: HERE

When a message cannot be processed correctly, it's not acknowledged. The message is then redelivered and probably again not processed correctly, which causes the cycle to execute indefinitely.

> To prevent a message that can't be properly managed from endlessly redelivering, configure the *Max redelivery* field to set a maximum number of times that a message can be delivered to the application.  +

Reviewer: ```suggestion
To prevent a message from endless redelivery, configure the *Max redelivery* field.  +
```
Name of the field tells what it is.

> By default, JMS Connector uses a maximum redelivery of `0`, which means that messages won't be redelivered, and doesn't matter if the message is recovered or rolled back from a transaction. If the message has persistent delivery set, ActiveMQ sends the message to a `DLQ.QUEUE` (dead letter queue).

Reviewer: Hmmm...seems to me that if the default is 0, then maybe we should start with this: that the default is 0, which means.... However, if the max delivery value is set to a value that is too high, then <this> happens, so if not using the default, take care with the value you do set.

Reviewer: Copyedits, existing text:
```suggestion
By default, JMS Connector uses a maximum redelivery value of `0`, which means that messages won't be redelivered, regardless of whether the message is recovered or rolled back from a transaction. If the message has persistent delivery set, ActiveMQ sends the message to `DLQ.QUEUE` (a dead letter queue).
```

> ActiveMQ enables you also to have a client-side redelivery configuration:

Reviewer: ```suggestion
ActiveMQ also enables you also to configure client-side redelivery.
```

> * The *Initial redelivery delay* field enables you to configure in milliseconds how much time to wait for the message to be redelivered the first time. +

Reviewer: ```suggestion
* The *Initial redelivery delay* field enables you to configure, in milliseconds, how much time to wait before the first message redelivery. +
```

> * The *Redelivery delay* field enables you to configure in milliseconds how much time to wait until the message is subsequently redelivered after the first redelivery.

Reviewer: ```suggestion
* The *Redelivery delay* field enables you to configure, in milliseconds, how much time to wait before subsequent redeliveries.
```

To configure these fields in Studio, follow these steps:

. In Studio, open the *JMS Config* global element window.
. In the *Connection Factory* section, set the *Factory configuration* field to *Edit inline*.
. Set the *Initial redelivery delay* field to `1000`.
. Set the *Redelivery delay* field to `100`.
. Set the *Max redelivery* field to `10`.
. Click *OK*.

.JMS Connector Redelivery configuration
image::jms-activemq-redelivery.png[Set the Max redelivery field to 10]

> In the XML editor, the `maxRedelivery`,`redeliveryDelay` and `initialRedeliveryDelay` configurations look like this:

Reviewer: ```suggestion
In the XML editor, the `maxRedelivery`,`redeliveryDelay`, and `initialRedeliveryDelay` configurations look like this:
```

[source,xml,linenums]
----
<jms:config name="JMS_Config">
  <jms:active-mq-connection>
    <jms:factory-configuration maxRedelivery="10"
      redeliveryDelay="100"
      initialRedeliveryDelay="1000"/>
  </jms:active-mq-connection>
</jms:config>
----

== Configure Trusted Packages

> 

Reviewer: HERE

> ActiveMQ versions 5.12.2, 5.13.0, and later restricts the classes that can be serialized and deserialized to prevent the execution of a malicious payload that can exploit the host system.

Reviewer: versions restrict
```suggestion
ActiveMQ versions 5.12.2, 5.13.0, and later restrict the classes that can be serialized and deserialized, helping to prevent the execution of a malicious payload on the host system.
```
"classes" of what?

> For ActiveMQ, JMS Connector enables only the JDK and JRE provided classes by default, so if you

Reviewer: ```suggestion
For ActiveMQ, JMS Connector enables by default only the JDK-provided and JRE-provided, so if you
```

need to exchange object messages, you must add the packages your applications use by activating the *Trusted packages* field. +

> Even though you can also enable the *Trust all packages* field to allow the serialization of more classes by allowing any object to be serialized and deserialized, this is not as secure as leaving it disabled in most cases. Keep this parameter disabled to improve the security and help prevent malicious attacks.

Reviewer: ```suggestion
Even though it is available to you, avoid enabling the *Trust all packages* field, which limits your ability to improve security and prevent malicious attacks.
```
I think that the previous text has explained about the danger of uncontrolled serialization/deserialization, so we can shorten this quite a bit.

In the following example, you configure the connection to enable only users who are compliant with the `com.mulesoft.someapp` and `com.mulesoft.someapp.model` packages to consume and produce ObjectMessages:

. In Studio, open the *JMS Config* global element window.
. In the *Connection Factory* section, set the *Factory configuration* field to *Edit inline*.
. Select the *Trusted packages* field.

> . Click the plus sign to add a new package.

Reviewer: ```suggestion
. Click *+* to add a new package.
```

. In the *Trusted package* window, set the *Value* field to `com.mulesoft.someapp`.
. Click *Finish*.
. Repeat Step 4.
. In the *Trusted package* window, set the *Value* field to `com.mulesoft.someapp.model`.
. Click *Finish*.

> . Select the *Trust all packages* field.

Reviewer: Wait...what? Didn't we just tell them to not use this field?

Reviewer: Maybe I don't understand something about the _combination_ of adding trusted packages and trusting all packages?
